### ISSUE TITLE
CMake Version Issue

### ISSUE THREAD

**Issue by jatinchowdhury18** [moddevices/mod-plugin-builder#111]
I've been working to compile one of my plugins for MOD Devices, and I've run into an issue with CMake versioning. The version of CMake currently used in this buildroot setup is version 3.4. However, my plugin needs (minimum) version 3.15.Is there a way to update the CMake version used by buildroot? Or maybe there is another workaround I could try?

Thanks,
Jatin

**Comment by falkTX** [comment]
We can update cmake, sure. We already did that before for sfizz, but that was just from 3.4 to 3.5
Any special reason you need 3.15 in particular?

**Comment by jatinchowdhury18** [comment]
That would be great! I need 3.15 since I'm using the [JUCE CMake API](https://github.com/juce-framework/JUCE/blob/master/docs/CMake%20API.md), which requires version 3.15 or higher. I wish I understood what they're doing internally that requires such a recent version, by my understanding of CMake isn't quite there yet.

**Comment by falkTX** [comment]
well, that won't work with raw/upstream juce since the MOD platform has no X11, so compilation will fail.

**Comment by jatinchowdhury18** [comment]
Ah thanks for clarifying...

I'm currently using a [fork of JUCE6](https://github.com/Chowdhury-DSP/DISTRHO-JUCE/tree/distrho-fork-lv2) based on the DISTRHO-JUCE fork, that I've modified to support LV2 plugin builds with CMake.

I had read on the [JUCE forum](https://forum.juce.com/t/x11-midi-and-other-linux-improvements/20945/18), that with JUCE6 the X11 code had been somewhat de-coupled from the general Linux UI code, so I had hoped this would make it easier to make truly "headless" plugin builds on Linux. That said, I'm not really sure about a lot of the details, plugin I'm not sure how the LV2 wrapper on my fork is handling the X11 dependency. Hopefully next weekend I'll have some time to experiment...

**Comment by falkTX** [comment]
X11 is not needed at runtime for JUCE anymore, so plugins can run really headless, but it is still needed at compile time.
I have a custom macro to disable all UI stuff on the old version of juce, but I have yet to make this work with the JUCE6 stuff.

Your plugins are all opensource right? I think we can work something out together to get this all building correctly.
Based on https://github.com/DISTRHO/DISTRHO-Ports/blob/master/libs/juce-current/patches/21_audioprocessor-no-gui.patch we can have JUCE with no UI stuff at all, but I have yet to test that with an actual plugin in new juce codebase.
Please try it, just define `JUCE_AUDIOPROCESSOR_NO_GUI` as compiler macro, and try to build as according to here https://github.com/DISTRHO/DISTRHO-Ports/blob/master/libs/juce-current/meson.build#L3 and https://github.com/DISTRHO/DISTRHO-Ports/blob/master/libs/juce-legacy/AppConfig.h#L500 (basically, skip enabling or building certain modules)

**Comment by jatinchowdhury18** [comment]
Okay, that makes sense.

Yes, the current plugin that I'm working to port is [this one](https://github.com/jatinchowdhury18/KlonCentaur), although I'm hoping to port some others as well once I get the process figured out!

Hopefully tonight or tomorrow I'll have a chance to try out this build process. Thanks for sharing the details!

**Comment by falkTX** [comment]
btw, best to target moddwarf first, and that is the one with the more recent gcc.
duo and duox still at 4.9 for keeping compatibility intact.

**Comment by jatinchowdhury18** [comment]
Okay, I finally got around to trying to build with no GUI, and just wanted to update with my latest progress. I was able to manually force buildroot to use a higher CMake version, and set the `JUCE_AUDIOPROCESSOR_NO_GUI` flag. However, I'm having some trouble with disabling the JUCE gui modules.

When JUCE builds through CMake, they have a sort of "utility" executable that gets built first called `juceaide`, and for some reason it seems like `juceaide` needs some of the functions from `juce_gui_basics`. Anyway, I'm still trying to figure out how I can get around this, either by not building `juceaide`, or by getting `juceaide` to work without the `juce_gui_basics` module.
